### PR TITLE
Multiplexer's connection to handle matching to support subclasses

### PR DESCRIPTION
**Pull request by mmandlis**
for: #842

### CODE REVIEW

File: runtime/particle.js [hunk 1]
```diff
@@ -177,6 +177,19 @@ export class Particle {
     }
     return false;
   }
+  _matchTypes(connection, handle) {
```
Comment (shaper) [inline runtime/particle.js]: It seems worth adding unit tests for this logic even though it's a private method. Is it a private method, in fact? It's used in `Multiplexer.js`. Maybe it shouldn't have the '_' prefix?

Comment (mmandlis) [reply runtime/particle.js]: refactored and added test. thanks!

**Comment by mmandlis** [comment]
PTAL

File: runtime/test/type-integration-test.js [hunk 2]
```diff
@@ -57,4 +57,17 @@ describe('type integration', () => {
...
+    let productType = manifest.findSchemaByName('Product').entityClass().type;
+    let legoType = manifest.findSchemaByName('Lego').entityClass().type;
+    assert.isTrue(productType.equalsOrSubclass(productType));
+    assert.isTrue(legoType.equalsOrSubclass(legoType));
+    assert.isTrue(legoType.equalsOrSubclass(productType));
+    assert.isTrue(legoType.setViewOf().equalsOrSubclass(productType.setViewOf()));
+
+    assert.isFalse(productType.equalsOrSubclass(legoType));
+    assert.isFalse(legoType.setViewOf().equalsOrSubclass(productType));
+  });
```
Comment (shaper) [inline runtime/test/type-integration-test.js]: Could be worth a case to validate false when comparing to a different type that is totally unrelated to `Product` and `Lego`.

File: shell/artifacts/Common/source/Multiplexer.js [hunk 3]
```diff
@@ -31,7 +31,7 @@ defineParticle(({TransformationDomParticle}) => {
           continue;
         }
         otherMappedViews.push(`map '${await arc.mapHandle(otherView._proxy)}' as v${index}`);
-        let hostedOtherConnection = hostedParticle.connections.find(conn => conn.type.equals(otherView.type));
+        let hostedOtherConnection = hostedParticle.connections.find(conn => otherView.type.equalsOrSubclass(conn.type));
```
Comment (dstockwell) [inline shell/artifacts/Common/source/Multiplexer.js]: This should use the type checker to determine whether the types are compatible, like this: https://github.com/PolymerLabs/arcs/blob/master/runtime/recipe/handle-connection.js#L115

File: runtime/test/type-integration-test.js (hunk 2 above)
Comment (mmandlis) [reply runtime/test/type-integration-test.js]: done

File: shell/artifacts/Common/source/Multiplexer.js (hunk 3 above)
Comment (mmandlis) [reply shell/artifacts/Common/source/Multiplexer.js]: Added a helper method to connection-spec, using the TypeChecker.compareTypes.
PTAL

File: runtime/recipe/type-checker.js [hunk 4]
```diff
@@ -113,8 +113,8 @@ class TypeChecker {
     // guarantees that the view's type will be preserved, and that the fact
     // that the type comes from a view rather than a connection will also
     // be preserved.
-    let superDirection = superclass.connection ? superclass.connection.direction : 'inout';
-    let subDirection = subclass.connection ? subclass.connection.direction : 'inout';
+    let superDirection = superclass.direction || (superclass.connection ? superclass.connection.direction : 'inout');
```
Comment (shaper) [inline runtime/recipe/type-checker.js]: Needs @dstockwell review as I'm too new to this part of the code to know whether this perhaps weakens the conditions the comment above this describes. It did prompt me to spend some time reading code re: connections.

File: runtime/test/recipe-test.js [hunk 5]
```diff
@@ -39,4 +39,59 @@ describe('recipe', function() {
...
   });
+  it('validate handle connection types', async () => {
+    let manifest = await Manifest.parse(`
+        schema MyType
+        schema MySubType extends MyType
+        schema OtherType
+        particle P1
+          P1(in MyType inMy)
+        particle P2
+          P2(out MyType inMy)
```
Comment (shaper) [inline runtime/test/recipe-test.js]: Rename to `outMy` just for clarity? Or just rename all of the particle params to something generic like `dummy` since it doesn't look like they really matter.

File: runtime/test/recipe-test.js [hunk 6]
```diff
@@ -39,4 +39,59 @@ describe('recipe', function() {
...
+        particle P3
+          P3(in MySubType inMy)
+        particle P4
+          P4(out MySubType inMy)
```
Comment (shaper) [inline runtime/test/recipe-test.js]: Rename to `outMy` just for clarity?

File: runtime/recipe/type-checker.js (hunk 4 above)
Comment (mmandlis) [reply runtime/recipe/type-checker.js]: absolutely, i'll wait for Doug's review.

File: runtime/particle-spec.js [hunk 7]
```diff
@@ -29,6 +30,10 @@ class ConnectionSpec {
   get isOutput() {
     return this.direction == 'out' || this.direction == 'inout';
   }
+
+  isValidType(type) {
```
[the remark this answers is not in the document]
Comment (mmandlis) [reply runtime/particle-spec.js]: renamed

File: runtime/test/recipe-test.js (hunk 5 above)
Comment (mmandlis) [reply runtime/test/recipe-test.js]: done

File: runtime/test/recipe-test.js (hunk 6 above)
Comment (mmandlis) [reply runtime/test/recipe-test.js]: done
